R-packages/modeltools/R/matrix.R

```diff
@@ -50,50 +65,106 @@
-                       train_end_date - training_window_size + 1,
-                       train_end_date)) %>%
-        select(tidyselect::starts_with(sprintf("response+%i:", ahead))) %>%
-        pull()
-
-    # Prediction matrices
-    out$predict_x <- lagged_df %>%
+create_train_and_predict_matrices <- function(lagged_df, ahead, training_window_size,
+                                              aheads_separate = TRUE) {
+    all_out <- list()
+    
+    # prediction matrices are the same for all aheads
+    predict_x <- lagged_df %>%
         filter(time_value == max(time_value)) %>%
         select(tidyselect::starts_with("value")) %>%
         as.matrix()
-    out$predict_geo_values <- lagged_df %>%
+    predict_geo_values <- lagged_df %>%
         filter(time_value == max(time_value)) %>%
         select(geo_value) %>% pull()
-    out$train_end_date <- train_end_date
-
-    return(out)
+    
+    # make sure the response columns are unique
+    for (a in ahead) {
+        responses_at_ahead <- lagged_df %>%
+            select(tidyselect::starts_with(sprintf("response+%i:", a))) %>%
+            ncol()
+        assert_that(responses_at_ahead == 1,
+                    msg = paste("multiple responses at ahead =", a))
+    }
+    
+    if (aheads_separate) {
+        for (a in ahead) {
+            out <- list()  # matrices corresponding to ahead+a
+            
+            # Find the last possible date of training data
+            response_end_date <- lagged_df %>%
+                select(time_value, tidyselect::starts_with(sprintf("response+%i:", a))) %>%
+                tidyr::drop_na() %>%
+                summarize(max(time_value)) %>%
+                pull()
+            train_end_date <- min(max(lagged_df$time_value), response_end_date)
+            
+            # Training matrices
+            out$train_x <- lagged_df %>%
+                select(geo_value, time_value, tidyselect::starts_with("value")) %>%
+                filter(between(time_value,
+                               train_end_date - training_window_size + 1,
+                               train_end_date)) %>%
+                select(-c(geo_value, time_value)) %>%
+                as.matrix()
+            out$train_y <- lagged_df %>%
+                filter(between(time_value,
+                               train_end_date - training_window_size + 1,
+                               train_end_date)) %>%
+                select(tidyselect::starts_with(sprintf("response+%i:", a))) %>%
+                pull()
+            
+            # Add prediction matrices and training end date
+            out$predict_x <- predict_x
+            out$predict_geo_values <- predict_geo_values
+            out$train_end_date <- train_end_date
+            
+            all_out[[paste0("ahead+", a)]] <- out
+        }
+        
+        if (length(ahead) == 1) {
+            return(all_out[[1]])
+        } else {
+            return(all_out)
+        }
+    } else {
```

Comment:
Let's note here what the `else` is doing explicitly: `! aheads_separate` or some such

Reply:
Done (in comment).